### PR TITLE
BUG: check CRS authority via SRID number when reading from postgis

### LINKED ISSUE #2545: BUG: CRS not found exception loading data frame
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of geopandas.

- [ ] (optional) I have confirmed this bug exists on the main branch of geopandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import psycopg2
import geopandas as gpd

con = psycopg2.connect(database="data", user="postgres", password="postgres", host="localhost")
sql = "SELECT *, public.ST_Transform(wkb_geometry, 54052) as geom FROM public.waypoint"
points = gpd.read_postgis(sql, con, geom_col='geom')
```

#### Problem description

I initially reported this bug [at pyproj](https://github.com/pyproj4/pyproj/issues/1126), but was then directed here. This issue 
could be caused by inappropriate use of some pyproj method(s).

My system is Ubuntu 22.04. The use case is to read into a data frame a set of
points from around the globe and plot it to map. The points are in a PostGIS
database. Proj fails with an exception when reading the points to a GeoPandas
data frame. It claims the CRS does not exist, but it does.

#### Full log

This is the full environment:

```
$ pip freeze
attrs==22.1.0
certifi==2022.6.15
click==8.1.3
click-plugins==1.1.1
cligj==0.7.2
Fiona==1.8.21
geopandas==0.11.1
munch==2.5.0
numpy==1.23.2
packaging==21.3
pandas==1.4.4
psycopg2==2.9.3
pyparsing==3.0.9
pyproj==3.3.1
python-dateutil==2.8.2
pytz==2022.2.1
Shapely==1.8.4
six==1.16.0
```

The log from the interactive command line with the exception:

```
$ python3
Python 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import psycopg2
>>> import geopandas as gpd
>>>
>>> con = psycopg2.connect(database="data", user="postgres", password="postgres",
...     host="localhost")
>>>
>>> sql = "SELECT *, public.ST_Transform(wkb_geometry, 54052) as geom FROM public.waypoint"
>>> points = gpd.read_postgis(sql, con, geom_col='geom')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/duque004/git/wrb-class/env/lib/python3.10/site-packages/geopandas/io/sql.py", line 163, in _read_postgis
    return _df_to_geodf(df, geom_col=geom_col, crs=crs)
  File "/home/duque004/git/wrb-class/env/lib/python3.10/site-packages/geopandas/io/sql.py", line 93, in _df_to_geodf
    return GeoDataFrame(df, crs=crs, geometry=geom_col)
  File "/home/duque004/git/wrb-class/env/lib/python3.10/site-packages/geopandas/geodataframe.py", line 189, in __init__
    self.set_geometry(geometry, inplace=True, crs=crs)
  File "/home/duque004/git/wrb-class/env/lib/python3.10/site-packages/geopandas/geodataframe.py", line 347, in set_geometry
    level = _ensure_geometry(level, crs=crs)
  File "/home/duque004/git/wrb-class/env/lib/python3.10/site-packages/geopandas/geodataframe.py", line 59, in _ensure_geometry
    out = from_shapely(np.asarray(data), crs=crs)
  File "/home/duque004/git/wrb-class/env/lib/python3.10/site-packages/geopandas/array.py", line 150, in from_shapely
    return GeometryArray(vectorized.from_shapely(data), crs=crs)
  File "/home/duque004/git/wrb-class/env/lib/python3.10/site-packages/geopandas/array.py", line 285, in __init__
    self.crs = crs
  File "/home/duque004/git/wrb-class/env/lib/python3.10/site-packages/geopandas/array.py", line 335, in crs
    self._crs = None if not value else CRS.from_user_input(value)
  File "/home/duque004/git/wrb-class/env/lib/python3.10/site-packages/pyproj/crs/crs.py", line 488, in from_user_input
    return cls(value, **kwargs)
  File "/home/duque004/git/wrb-class/env/lib/python3.10/site-packages/pyproj/crs/crs.py", line 335, in __init__
    self._local.crs = _CRS(self.srs)
  File "pyproj/_crs.pyx", line 2352, in pyproj._crs._CRS.__init__
pyproj.exceptions.CRSError: Invalid projection: epsg:54052: (Internal Proj Error: proj_create: crs not found)
```
 
First check, the CRS definition in PostGIS:

```
$ psql -d data
psql (14.5 (Ubuntu 14.5-0ubuntu0.22.04.1))
Type "help" for help.

data=# select * from public.spatial_ref_sys where srid = 54052;

 srid  | auth_name | auth_srid |                                                                                                                                                                                                                                srtext                                                                                                                                                                                                                                |                             proj4text
-------+-----------+-----------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------
 54052 | ESRI      |     54052 | PROJCS["World_Goode_Homolosine_Land",GEOGCS["WGS 84",DATUM["WGS_1984",SPHEROID["WGS 84",6378137,298.257223563,AUTHORITY["EPSG","7030"]],AUTHORITY["EPSG","6326"]],PRIMEM["Greenwich",0],UNIT["Degree",0.0174532925199433]],PROJECTION["Goode_Homolosine"],PARAMETER["central_meridian",0],PARAMETER["false_easting",0],PARAMETER["false_northing",0],UNIT["metre",1,AUTHORITY["EPSG","9001"]],AXIS["Easting",EAST],AXIS["Northing",NORTH],AUTHORITY["ESRI","54052"]] | +proj=goode +lon_0=0 +x_0=0 +y_0=0 +datum=WGS84 +units=m +no_defs
(1 row)
```

Then check the same in the Proj database:


```
$ sqlite3 ./lib/python3.10/site-packages/pyproj/proj_dir/share/proj/proj.db
SQLite version 3.37.2 2022-01-06 13:25:41
Enter ".help" for usage hints.
sqlite> select * from projected_crs where code = 54052;
ESRI|54052|World_Goode_Homolosine_Land||||EPSG|4326|||PROJCS["World_Goode_Homolosine_Land",GEOGCS["GCS_WGS_1984",DATUM["D_WGS_1984",SPHEROID["WGS_1984",6378137.0,298.257223563]],PRIMEM["Greenwich",0.0],UNIT["Degree",0.0174532925199433]],PROJECTION["Goode_Homolosine"],PARAMETER["False_Easting",0.0],PARAMETER["False_Northing",0.0],PARAMETER["Central_Meridian",0.0],PARAMETER["Option",1.0],UNIT["Meter",1.0]]|0
```

Apparently, GeoPandas is inadvertently changing the CRS authority, but this might
not explain the exception. The `srid` field in PostGIS and `code` in the Proj
database are both unique (not enforced, but the codes do not repeat). The programme 
should be able to fetch the correct CRS definition with the code alone.

Thank you for reading.


#### Expected Output

The points layer should be loaded as a data frame with no exception.

#### Output of ``geopandas.show_versions()``

<details>


SYSTEM INFO
-----------
python     : 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0]
executable : /home/duque004/git/wrb-class/env/bin/python
machine    : Linux-5.15.0-46-generic-x86_64-with-glibc2.35

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : 3.10.2
GEOS lib   : /usr/lib/x86_64-linux-gnu/libgeos_c.so
GDAL       : 3.4.1
GDAL data dir: /home/duque004/git/wrb-class/env/lib/python3.10/site-packages/fiona/gdal_data
PROJ       : 8.2.0
PROJ data dir: /home/duque004/git/wrb-class/env/lib/python3.10/site-packages/pyproj/proj_dir/share/proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.11.1
numpy      : 1.23.2
pandas     : 1.4.4
pyproj     : 3.3.1
shapely    : 1.8.4
fiona      : 1.8.21
geoalchemy2: None
geopy      : 2.2.0
matplotlib : 3.5.3
mapclassify: 2.4.3
pygeos     : None
pyogrio    : None
psycopg2   : 2.9.3 (dt dec pq3 ext lo64)
pyarrow    : None
rtree      : None


</details>

### PR DESCRIPTION
**Pull request by nicholas-ys-tan**
Should resolve #2545 

It looks at the `spatial_ref_sys` table in postgis to look up the auth name instead of assuming EPSG. However, if for whatever reason the table does not exist, it should default back to current  behaviour of assuming epsg.

Not sure if this is the best way though, I didn't want to pass the connection data into `_df_to_geodf`, so read it into a dataframe first. Would be nice if such a look up for the authority name could be done purely in `pyproj.CRS` but I couldn't find anything.

Please advise if there is a more aesthetic approach to this.

### CODE REVIEW

File: geopandas/io/sql.py [hunk 1]
```diff
@@ -92,8 +96,11 @@ def load_geom_text(x):
             srid = shapely.get_srid(geoms.iat[0])
             # if no defined SRID in geodatabase, returns SRID of 0
             if srid != 0:
-                crs = "epsg:{}".format(srid)
-
+                if spatial_ref_sys_df is not None:
+                    entry = spatial_ref_sys_df.loc[spatial_ref_sys_df["srid"] == srid]
+                    crs = f"{entry['auth_name'].values[0]}:{srid}"
```
Comment (m-richards) [inline geopandas/io/sql.py:101]:
```suggestion
                    crs = f"{entry['auth_name'].item()}:{srid}"
```
I would expect this should be a unique match (although perhaps if there are somehow multiple, falling back to the first is okay)

**Review by m-richards — APPROVED** [review]
Thanks @nicholas-ys-tan, this approach looks sound to me. To me it makes sense that we treat postgis rather than the proj.db as the point of truth for data coming out of postgis. I think passing in the dataframe rather than the connection makes sense - especially for the chunksize case as it means this is only queried once upfront.

File: geopandas/io/sql.py [hunk 2]
```diff
@@ -92,8 +96,11 @@ def load_geom_text(x):
             srid = shapely.get_srid(geoms.iat[0])
             # if no defined SRID in geodatabase, returns SRID of 0
             if srid != 0:
-                crs = "epsg:{}".format(srid)
-
+                if spatial_ref_sys_df is not None:
+                    entry = spatial_ref_sys_df.loc[spatial_ref_sys_df["srid"] == srid]
```
[the remark this answers is not in the document]
Comment (nicholas-ys-tan) [reply geopandas/io/sql.py:102]: Thanks @m-richards ,

I've added an additional check that the `srid` is in the `spatial_ref_sys` table, otherwise falling back to epsg and added a user warning when it does this. I've written a test to check this using `mock.patch`

File: geopandas/io/sql.py [hunk 3]
```diff
@@ -156,6 +173,14 @@ def _read_postgis(
     >>> df = geopandas.read_postgis(sql, con)  # doctest: +SKIP
     """
 
+    try:
+        spatial_ref_sys_sql = "SELECT * FROM spatial_ref_sys"
+        spatial_ref_sys_df = (
+            pd.read_sql(spatial_ref_sys_sql, con) if crs is None else None
+        )
```
Comment (jorisvandenbossche) [inline geopandas/io/sql.py:180]: Do you have an idea how large this table typically is? Because we are now reading this full table every time when calling `read_postgis`, and _ideally_ we would already do the filtering step here in the query, instead of doing it on the dataframe. I know we don't know the `srid` value at this point for specifying it in the query, but I wonder if we could maybe pass `con` to `_df_to_geodf` and then do this query there with directly filtering on srid. We would then only have to "cache" this somehow for repeated calls to `_df_to_geodf` from a single `read_postgis` call.

Comment (nicholas-ys-tan) [reply geopandas/io/sql.py:180]: Mine is 8000 rows with approx 7.3mb

```
postgres=# SELECT pg_size_pretty( pg_total_relation_size('spatial_ref_sys') );
 pg_size_pretty 
----------------
 7280 kB
(1 row)
```

Comment (jorisvandenbossche) [reply geopandas/io/sql.py:180]: That's a decent chunk of data to have to transfer each time. I would try to look into querying only the one `srid` value that is needed.

File: geopandas/io/sql.py [hunk 4]
```diff
@@ -92,7 +95,35 @@ def load_geom_text(x):
...
             if srid != 0:
-                crs = "epsg:{}".format(srid)
+                try:
+                    spatial_ref_sys_df = _get_spatial_ref_sys_df(con, srid)
+
+                    spatial_ref_sys_sql = (
+                        f"SELECT srid, auth_name FROM "
+                        f"spatial_ref_sys WHERE srid = {srid}"
+                    )
+                    spatial_ref_sys_df = pd.read_sql(spatial_ref_sys_sql, con)
```
Comment (m-richards) [inline geopandas/io/sql.py:105]:
```suggestion
```
this bit looks duplicated right now

Comment (nicholas-ys-tan) [reply geopandas/io/sql.py:105]: whoops, what am embarrassing thing to miss, thanks!

File: geopandas/io/.sql.py.swp
Comment (jorisvandenbossche) [inline geopandas/io/.sql.py.swp:1]: This file was added by accident?

File: geopandas/io/tests/test_sql.py [hunk 5]
```diff
@@ -804,3 +810,48 @@ def test_duplicate_geometry_column_fails(self, engine_postgis):
...
+        df_nybb = df_nybb.to_crs(crs="esri:54052")
+        create_postgis(con, df_nybb, srid=54052)
+
+        sql = "SELECT * FROM nybb;"
+        df = read_postgis(sql, con)
+        validate_boro_df(df)
+        assert df.crs == "ESRI:54052"
+
+    @pytest.mark.skipif(not HAS_PYPROJ, reason="pyproj not installed")
+    @mock.patch("shapely.get_srid")
```
Comment (jorisvandenbossche) [inline geopandas/io/tests/test_sql.py:826]: Maybe could also add a test mocking `_get_spatial_ref_sys_df` to raise an error or to return an empty dataframe

**Comment by jorisvandenbossche** [comment]
Can you also update with latest main to get the CI fix from #3339?

File: geopandas/io/.sql.py.swp
Comment (nicholas-ys-tan) [reply geopandas/io/.sql.py.swp:1]: apologies, missed that, deleted

File: geopandas/io/sql.py [hunk 6]
```diff
@@ -92,7 +95,29 @@ def load_geom_text(x):
...
+                    warning_msg = (
+                        f"Could not find the spatial reference system table "
+                        f"(spatial_ref_sys) in PostGIS."
+                        f"Trying epsg:{srid} as a fallback."
+                    )
+                    spatial_ref_sys_df = None
+                    warnings.warn(warning_msg, UserWarning, stacklevel=2)
+                    crs = "epsg:{}".format(srid)
+
+                if not spatial_ref_sys_df.empty:
```
[the remark this answers is not in the document]
Comment (nicholas-ys-tan) [reply geopandas/io/sql.py]: yes, I missed that would happen

added `else` and removed the consequent redundant `spatial_ref_sys_df = None`

**Comment by nicholas-ys-tan** [comment]
> Can you also update with latest main to get the CI fix from #3339?

Rebased and seems to have resolved the engine issues in the windows build but seem to still have one more from a code coverage issue in one of the windows CIs, is that from my end?

**Comment by martinfleis** [comment]
> is this from my end?

ignore that.

File: geopandas/io/sql.py [hunk 7]
```diff
@@ -92,7 +95,28 @@ def load_geom_text(x):
...
-                crs = "epsg:{}".format(srid)
+                try:
+                    spatial_ref_sys_df = _get_spatial_ref_sys_df(con, srid)
+                except pd.errors.DatabaseError:
+                    warning_msg = (
+                        f"Could not find the spatial reference system table "
+                        f"(spatial_ref_sys) in PostGIS."
+                        f"Trying epsg:{srid} as a fallback."
+                    )
+                    warnings.warn(warning_msg, UserWarning, stacklevel=2)
```
Comment (jorisvandenbossche) [inline geopandas/io/sql.py:106]:
```suggestion
                    warnings.warn(warning_msg, UserWarning, stacklevel=3)
```

File: geopandas/io/sql.py [hunk 8]
```diff
@@ -92,7 +95,28 @@ def load_geom_text(x):
...
+                    if not spatial_ref_sys_df.empty:
+                        auth_name = spatial_ref_sys_df["auth_name"].item()
+                        crs = f"{auth_name}:{srid}"
+                    else:
+                        warning_msg = (
+                            f"Could not find srid {srid} in the "
+                            f"spatial_ref_sys table. "
+                            f"Trying epsg:{srid} as a fallback."
+                        )
+                        warnings.warn(warning_msg, UserWarning, stacklevel=2)
```
Comment (jorisvandenbossche) [inline geopandas/io/sql.py:118]:
```suggestion
                        warnings.warn(warning_msg, UserWarning, stacklevel=3)
```

**Review by jorisvandenbossche — APPROVED** [review]
Thanks for the updates, looks good!
